### PR TITLE
Feature/bartender expanded

### PR DESCRIPTION
**Pull request by MonkfishFillet**

### CODE REVIEW

**Comment by MonkfishFillet** [comment]
Plz review....

**Comment by DonChalupa** [comment]
This should be merged into develop, not main. I would be fine with merging straight to main if it included automated tests that cover all the edge cases and points of failure. Otherwise, our workflow is to branch from & merge to develop, perform testing ourselves, and hopefully uncover any bugs through deliberate testing or regular use as we continue to develop. 

Really, we should have automated tests.

File: cogs/Bartender.py [hunk 1]
```diff
@@ -1,25 +1,70 @@
...
+                filter = f'filter.php?i={arg}'
+                resolved_url = self.drink_url_factorty(drink_url, filter)
             
+            json = req.get(resolved_url).json()
+            drink = Drink(json['drinks'][0])
+            await ctx.send(embed = drink.embed)
+
         except Exception as ex:
             traceback.print_exc()
             await ctx.send(f'Ayo, your code is wack.\n Error: {ex}')
```
Comment (DonChalupa) [inline cogs/Bartender.py]: Handling potential exceptions from the API calls is a good idea, but doing so with one massive try/catch block that includes all of the work in this command is not. There should be deliberate try/catch calls on the API calls to handle that specific exception, as well as deliberate try/catch blocks for other statements that may fail.

Putting everything in one try/catch block makes it unclear exactly what you think could fail. Also, as this code grows or changes, that ambiguity makes it hard for another developer to maintain what you've got here while adding their own stuff. This would likely lead to a situation where they say "I don't know exactly how many exceptions this block is meant to catch so I'm just going to put my stuff under it to be safe". At best it's lazy, at worst it shows you don't actually know what to expect from the code.

File: cogs/Bartender.py [hunk 2]
```diff
@@ -1,25 +1,70 @@
...
+    @commands.command(name='drink', description=DRINK_DESCRIPTION, usage=DRINK_USAGE)
+    async def drink(self, ctx, *args):
+        FILTER_TYPES = ['a','bourbon','brandy','cognac','gin', 'na','rum','scotch','tequila','vodka','whiskey','wine']
+        drink_url = 'https://www.thecocktaildb.com/api/json/v1/1/'
+
         try:
-            json = req.get(DRINK_URL).json()['drinks'][0]
-            drink = Drink(json)
-            await ctx.send(embed = drink.embed)
+            if not args:
```
Comment (DonChalupa) [inline cogs/Bartender.py]: In these different cases of number of args, it would be more readable if this function, `drink`, tested these different cases, then called helper functions that describe and perform the relevant action for each case, rather than doing both here. 

E.g.,
```
if not args:
    return get_random_drink(...)
```

The intent is clearer, and it's simple to follow the function to know what's actually happening.

File: cogs/Bartender.py [hunk 3]
```diff
@@ -1,25 +1,70 @@
...
+                drink_url += 'random.php'
```
Comment (DonChalupa) [inline cogs/Bartender.py]: A minor thing, but it's safer to create a new variable for the random URL than to modify the `drink_url` variable.

```random_drink_url = drink_url + 'random.php'```

Suppose this code is changed later, and through either a deliberate change or an oversight this code runs when it shouldn't, or it fails to return. Now, the rest of the function operates with a modified `drink_url`, which may be incorrect or unexpected.

Note that this wouldn't be a concern if this code was refactored into a function as I suggested above.

File: cogs/Bartender.py [hunk 4]
```diff
@@ -1,25 +1,70 @@
...
+                json = req.get(drink_url).json()
+                drink = Drink(json['drinks'][0])
+                await ctx.send(embed = drink.embed)
+                return
+            
+
+            if len(args) > 1:
+                await ctx.send(f'I think you added too many arguments there, bud.')
```
Comment (DonChalupa) [inline cogs/Bartender.py]: Would be more helpful to instruct proper usage.

File: cogs/Bartender.py [hunk 5]
```diff
@@ -1,25 +1,70 @@
...
+                return
+
+            arg = args[0].lower()
+            if arg not in FILTER_TYPES:
+                    message = 'Better check your arguments there, bud.'
```
Comment (DonChalupa) [inline cogs/Bartender.py]: Again would be more helpful to instruct proper usage in this case.

File: cogs/Bartender.py [hunk 6]
```diff
@@ -1,25 +1,70 @@
...
+                    chance = r.randint(1,100)
+                    if chance > 80:
+                        message += ' (Kinda cringe...)'
+                    await ctx.send(message)
+                    return
+            
+            
+            if arg == 'a':
```
Comment (DonChalupa) [inline cogs/Bartender.py]: I would think 'Alcoholic' should also match for this case. Same with 'Non-alcoholic' (and variations) for the 'na' case.

File: cogs/Bartender.py [hunk 7]
```diff
@@ -1,25 +1,70 @@
...
+            drink = Drink(json['drinks'][0])
+            await ctx.send(embed = drink.embed)
+
         except Exception as ex:
             traceback.print_exc()
             await ctx.send(f'Ayo, your code is wack.\n Error: {ex}')
-    
+
+
+    def drink_url_factorty(self, drink_url, filter):
```
Comment (DonChalupa) [inline cogs/Bartender.py]: 1. "factory" is spelled incorrectly
2. This isn't really a 'factory' so to speak (as in the factory pattern). The factory pattern involves creating an instance of a polymorphic type at run-time. This is kind of just a normal function that gets a random drink given some filter. 
3. Continuing off the last point, the actual behavior of this function isn't described by its name. You aren't just constructing a URL from the given filter, you're finding the URL for a random drink given this filter. Something like `random_drink_url_from_filter(...)` is more appropriate (though admittedly more verbose - you could maybe include the next step in the process where you construct a drink object from this URL and call it `random_drink_from_filter(...)`, or make that part a helper function and include it inside this function).

File: cogs/Bartender.py [hunk 8]
```diff
@@ -1,25 +1,70 @@
...
+        filter_url = drink_url + filter
+        drink_list = req.get(filter_url).json()['drinks']
```
Comment (DonChalupa) [inline cogs/Bartender.py]: This can fail / throw an exception, so it should be wrapped in a try/except block.

**Comment by DonChalupa** [comment]
Honestly, this felt a little lower in quality than what I'm used to from you, like you got it working late one night and then just turned and made a pull request. Please make sure you're doing a quick review yourself before submitting. Things like spelling mistakes shouldn't really make it to this stage.

**Review by DonChalupa — CHANGES_REQUESTED** [review]

File: cogs/Bartender.py (hunk 1 above)
Comment (MonkfishFillet) [reply cogs/Bartender.py]: Jokes on you, I've always been this bad at coding llolololol

Comment (MonkfishFillet) [reply cogs/Bartender.py]: Refactored
https://github.com/DonChalupa/Robbot/commit/2b02393314efb1d65289e7c5c697dcbffcb6379d#diff-a1132eee6c43b832e1b4945ce331759b2cdb8e84dbe814c5cc4e4a31020ec7b1L57

File: cogs/Bartender.py (hunk 2 above)
Comment (MonkfishFillet) [reply cogs/Bartender.py]: I accidentally resolved the above conflict, however the following commit houses both fixes - please use this request for any further comments.

Though it's not in a function, I feel as if this structure sufficiently showcases the intent clearly while also reduces the redundancy of the request code.
https://github.com/DonChalupa/Robbot/commit/2b02393314efb1d65289e7c5c697dcbffcb6379d#diff-a1132eee6c43b832e1b4945ce331759b2cdb8e84dbe814c5cc4e4a31020ec7b1R21

File: cogs/Bartender.py (hunk 3 above)
Comment (MonkfishFillet) [reply cogs/Bartender.py]: Helper function created as well as a general get_drink function in order to reduce request code redundancy:
https://github.com/DonChalupa/Robbot/pull/16/commits/2b02393314efb1d65289e7c5c697dcbffcb6379d#diff-a1132eee6c43b832e1b4945ce331759b2cdb8e84dbe814c5cc4e4a31020ec7b1R43

File: cogs/Bartender.py (hunk 4 above)
Comment (MonkfishFillet) [reply cogs/Bartender.py]: Both argument validations have been refactored to show more detailed information:
https://github.com/DonChalupa/Robbot/pull/16/commits/2b02393314efb1d65289e7c5c697dcbffcb6379d#diff-a1132eee6c43b832e1b4945ce331759b2cdb8e84dbe814c5cc4e4a31020ec7b1R26

File: cogs/Bartender.py (hunk 5 above)
Comment (MonkfishFillet) [reply cogs/Bartender.py]: See above:
https://github.com/DonChalupa/Robbot/pull/16/commits/2b02393314efb1d65289e7c5c697dcbffcb6379d#diff-a1132eee6c43b832e1b4945ce331759b2cdb8e84dbe814c5cc4e4a31020ec7b1R35

File: cogs/Bartender.py (hunk 6 above)
Comment (MonkfishFillet) [reply cogs/Bartender.py]: Added to both the conditional logic and the allowed list of args:
https://github.com/DonChalupa/Robbot/pull/16/commits/2b02393314efb1d65289e7c5c697dcbffcb6379d#diff-a1132eee6c43b832e1b4945ce331759b2cdb8e84dbe814c5cc4e4a31020ec7b1R16
https://github.com/DonChalupa/Robbot/pull/16/commits/2b02393314efb1d65289e7c5c697dcbffcb6379d#diff-a1132eee6c43b832e1b4945ce331759b2cdb8e84dbe814c5cc4e4a31020ec7b1R56-R57

File: cogs/Bartender.py (hunk 7 above)
Comment (MonkfishFillet) [reply cogs/Bartender.py]: Refactored to reduce confusion:
https://github.com/DonChalupa/Robbot/pull/16/commits/2b02393314efb1d65289e7c5c697dcbffcb6379d#diff-a1132eee6c43b832e1b4945ce331759b2cdb8e84dbe814c5cc4e4a31020ec7b1R63

File: cogs/Bartender.py (hunk 8 above)
Comment (MonkfishFillet) [reply cogs/Bartender.py]: Abstracted out into its own function:
https://github.com/DonChalupa/Robbot/pull/16/commits/2b02393314efb1d65289e7c5c697dcbffcb6379d#diff-a1132eee6c43b832e1b4945ce331759b2cdb8e84dbe814c5cc4e4a31020ec7b1R43

File: robbot.py [hunk 9]
```diff
@@ -12,8 +12,8 @@
 CFG_FILENAME = 'config.json'
 
 load_dotenv()
-TOKEN = os.getenv('DISCORD_TOKEN')
-#TOKEN = os.getenv('LOCAL_TOKEN')
+#TOKEN = os.getenv('DISCORD_TOKEN')
+TOKEN = os.getenv('LOCAL_TOKEN')
```
Comment (DonChalupa) [inline robbot.py]: Make sure token is correct, looks like you swapped it here for the test Robbot

Comment (MonkfishFillet) [reply robbot.py]: @DonChalupa  https://github.com/DonChalupa/Robbot/pull/16/commits/6980b9adef711915524cca7a1bff50ed7eea2639

File: cogs/Bartender.py (hunk 8 above)
Comment (DonChalupa) [reply cogs/Bartender.py]: Still not resolved - think you were talking about the `req.get()` call on line 45 above, but this call (line 65) still needs to handle exceptions.

File: cogs/Bartender.py [hunk 10]
```diff
@@ -1,25 +1,72 @@
...
+ 
+        if not args:
+            random_drink_url = DRINK_URL + 'random.php'
+            drink = await self.get_drink(ctx, random_drink_url)
+            await ctx.send(embed=drink.embed)
+            return
             
-        except Exception as ex:
+        if len(args) > 1:
+            await ctx.send(f'There can only be one argument allow for this command, bud.')
```
Comment (DonChalupa) [inline cogs/Bartender.py:26]:
```suggestion
            await ctx.send(f'There can only be one argument allowed for this command, bud.')
```
That sentence is kinda wacky tbh, but not a big deal.

**Review by DonChalupa — CHANGES_REQUESTED** [review]
